packages/puppeteer-core/src/bidi/Page.ts

```diff
@@ -7,201 +7,93 @@
-    this.#browserContext = browserContext;
-    this.#connection = browsingContext.connection;
+  static from(
+    browserContext: BidiBrowserContext,
+    browsingContext: BrowsingContext
+  ): BidiPage {
+    const page = new BidiPage(browserContext, browsingContext);
+    page.#initialize();
+    return page;
+  }
 
-    for (const [event, subscriber] of this.#browsingContextEvents) {
-      this.#browsingContext.on(event, subscriber);
-    }
+  readonly #browserContext: BidiBrowserContext;
+  readonly #frame: BidiFrame;
+  #viewport: Viewport | null = null;
 
-    this.#networkManager = new BidiNetworkManager(this.#connection, this);
+  readonly keyboard: BidiKeyboard;
+  readonly mouse: BidiMouse;
+  readonly touchscreen: BidiTouchscreen;
+  readonly accessibility: Accessibility;
+  readonly tracing: Tracing;
+  readonly coverage: Coverage;
+  readonly #cdpEmulationManager: EmulationManager;
 
-    for (const [event, subscriber] of this.#subscribedEvents) {
-      this.#connection.on(event, subscriber);
-    }
-
-    for (const [event, subscriber] of this.#networkManagerEvents) {
-      // TODO: remove any
-      this.#networkManager.on(event, subscriber as any);
-    }
+  _client(): BidiCdpSession {
+    return this.#frame.client;
+  }
 
-    const frame = new BidiFrame(
-      this,
-      this.#browsingContext,
-      this._timeoutSettings,
-      this.#browsingContext.parent
-    );
-    this.#frameTree.addFrame(frame);
-    this.emit(PageEvent.FrameAttached, frame);
+  private constructor(
+    browserContext: BidiBrowserContext,
+    browsingContext: BrowsingContext
+  ) {
+    super();
+    this.#browserContext = browserContext;
+    this.#frame = BidiFrame.from(this, browsingContext);
 
-    // TODO: https://github.com/w3c/webdriver-bidi/issues/443
-    this.#accessibility = new Accessibility(
-      this.mainFrame().context().cdpSession
-    );
-    this.#tracing = new Tracing(this.mainFrame().context().cdpSession);
-    this.#coverage = new Coverage(this.mainFrame().context().cdpSession);
-    this.#cdpEmulationManager = new CdpEmulationManager(
-      this.mainFrame().context().cdpSession
-    );
-    this.#emulationManager = new EmulationManager(browsingContext);
-    this.#mouse = new BidiMouse(this.mainFrame().context());
-    this.#touchscreen = new BidiTouchscreen(this.mainFrame().context());
-    this.#keyboard = new BidiKeyboard(this);
+    this.#cdpEmulationManager = new EmulationManager(this.#frame.client);
+    this.accessibility = new Accessibility(this.#frame.client);
+    this.tracing = new Tracing(this.#frame.client);
+    this.coverage = new Coverage(this.#frame.client);
+    this.keyboard = new BidiKeyboard(this);
+    this.mouse = new BidiMouse(this);
+    this.touchscreen = new BidiTouchscreen(this);
   }
 
-  /**
-   * @internal
-   */
-  get connection(): BidiConnection {
-    return this.#connection;
+  #initialize() {
+    this.#frame.browsingContext.on('closed', () => {
```

Comment:
```suggestion
    this.#frame.browsingContext.once('closed', () => {
```

Reply:
This is not necessary because the browsing context removes all listeners after this event is emitted. Same for every other `bidi/core` object.